### PR TITLE
WIP: Stats email v2

### PR DESCRIPTION
**Pull request by oogetyboogety**
##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

### CODE REVIEW

**Comment by oogetyboogety** [comment]
@owocki This just need the hardcoded text to be localized. Otherwise, this should be finished, unless there are CSS changes to be included?

**Comment by oogetyboogety** [comment]
funder_email

**Comment by thelostone-mc** [comment]
Ah looks like something went wrong during your rebase. This PR has commits pushed in  by other folks ! Could you cherry-pick them and ensure the PR just has your commits ? ^_^

**Comment by oogetyboogety** [comment]
@thelostone-mc Done! Thanks, please provide any further instructions I can complete to get this merged!

Thanks again

**Comment by owocki** [comment]
looking at this now

File: app/dashboard/models.py [hunk 1]
```diff
@@ -1127,22 +1127,97 @@ def get_quarterly_stats(self):
...
+        user_fulfilled_bounties = False
+        user_funded_bounties = False
         last_quarter = datetime.now() - timedelta(days=90)
         bounties = self.bounties.filter(modified_on__gte=last_quarter)
         fulfilled_bounties = [
             bounty for bounty in bounties if bounty.is_hunter(self.handle) and bounty.status == 'done'
         ]
         fulfilled_bounties_count = len(fulfilled_bounties)
+        funded_bounties = [
+            bounty for bounty in bounties if bounty.is_funder(self.handle)
```
Comment (owocki) [inline app/dashboard/models.py]: is any of this data STUBBed out?  are you unsure of any of the calcs you did?  if so, mind throwing a few TODOs in there for us to work through?

File: app/dashboard/models.py [hunk 2]
```diff
@@ -1165,13 +1240,39 @@ def get_quarterly_stats(self):
...
+            relevant_bounties = []
+        else:
+            from marketing.utils import get_or_save_email_subscriber
+            user_coding_languages = get_or_save_email_subscriber(self.email, 'internal').keywords
+            if user_coding_languages is not None:
+                potential_bounties = Bounty.objects.all()
+                relevant_bounties = Bounty.objects.none()
+                for keyword in user_coding_languages:
+                    relevant_bounties = relevant_bounties.union(potential_bounties.filter(
+                            network='rinkeby',
```
Comment (owocki) [inline app/dashboard/models.py]: should not hardcode this network

**Comment by owocki** [comment]
over all looking good! heres my testing notes

1. double @ signs on this module http://bits.owocki.com/3Y0b100V2W28/Screen%20Shot%202018-06-04%20at%206.10.39%20PM.png
2. rounding http://bits.owocki.com/1o2b0D3H0J2d/Screen%20Shot%202018-06-04%20at%206.11.18%20PM.png

could you post screenshots from each of the personas pls?  that'll make testing/QA easier for everyone else

cc @mbeacom @thelostone-mc in case anyone else wants to QA too

**Comment by oogetyboogety** [comment]
@owocki Done, please review!

File: app/dashboard/models.py [hunk 3]
```diff
@@ -1127,22 +1090,97 @@ def get_quarterly_stats(self):
             dict : containing the following information
             'user_total_earned_eth': Total earnings of user in ETH.
             'user_total_earned_usd': Total earnings of user in USD.
-            'user_fulfilled_bounties_count': Total bounties fulfilled by user.
+            'user_total_funded_usd': Total value of bounties funded by the user on bounties in done status in USD
+            'user_total_funded_hours': Total hours input by the developers on the fulfillment of bounties created by the user in USD
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/models.py]: ^

File: app/dashboard/models.py [hunk 4]
```diff
@@ -1127,22 +1090,97 @@ def get_quarterly_stats(self):
...
             'user_avg_eth_earned_per_bounty': Average earning in ETH earned by user per bounty
             'user_avg_usd_earned_per_bounty': Average earning in USD earned by user per bounty
             'user_num_completed_bounties': Total no. of bounties completed.
+            'user_num_funded_fulfilled_bounties': Total bounites that were funded by the user and fulfilled
             'user_bounty_completion_percentage': Percentage of bounties successfully completed by the user
+            'user_funded_fulfilled_percentage': Percentage of bounties funded by the user that were fulfilled
             'user_active_in_last_quarter': bool, if the user was active in last quarter
             'user_no_of_languages': No of languages user used while working on bounties.
             'user_languages': Languages that were used in bounties that were worked on.
+            'relevant_bounties': a list of Bounty(s) that would match the skillset input by the user into the Match tab of their settings
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/models.py]: ^

File: app/dashboard/models.py [hunk 5]
```diff
@@ -1127,22 +1127,100 @@ def get_quarterly_stats(self):
...
             bounty for bounty in bounties if bounty.is_hunter(self.handle) and bounty.status == 'done'
         ]
         fulfilled_bounties_count = len(fulfilled_bounties)
+        funded_bounties = [
+            bounty for bounty in bounties if bounty.is_funder(self.handle)
+        ]
+        funded_bounties_count = len(funded_bounties)
+        if funded_bounties_count:
+            total_funded_usd = sum([
+                bounty.value_in_usdt if bounty.value_in_usdt else 0
```
Comment (mbeacom) [inline app/dashboard/models.py]: We should use [Django Aggregation](https://docs.djangoproject.com/en/2.0/topics/db/aggregation/) here.

File: app/dashboard/models.py [hunk 6]
```diff
@@ -1127,22 +1127,97 @@ def get_quarterly_stats(self):
...
+                    if isinstance(fulfillment.fulfiller_hours_worked, Decimal):
+                        total_funded_hours += fulfillment.fulfiller_hours_worked
+                        funded_fulfillments_with_hours_counted += 1
+
+            user_funded_bounty_developers = []
+            for fulfillment in funded_bounty_fulfillments:
+                user_funded_bounty_developers.append(fulfillment.fulfiller_github_username.lstrip('@'))
+            user_funded_bounty_developers = [*{*user_funded_bounty_developers}]
+            if funded_fulfillments_with_hours_counted:
+                avg_hourly_rate_per_funded_bounty = float(total_funded_hourly_rate) / float(funded_fulfillments_with_hours_counted)
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/models.py]: ^

File: app/dashboard/models.py [hunk 7]
```diff
@@ -1165,13 +1243,41 @@ def get_quarterly_stats(self):
...
+                potential_bounties = Bounty.objects.all()
+                relevant_bounties = Bounty.objects.none()
+                for keyword in user_coding_languages:
+                    relevant_bounties = relevant_bounties.union(potential_bounties.filter(
+                            network=Profile.get_network(),
+                            current_bounty=True,
+                            metadata__icontains=keyword,
+                            idx_status__in=['open'],
+                            #TODO: Identify the random sampling method used, see if it sending mass emails becomes slow
+                            ).order_by('?')
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/models.py]: ^

File: app/retail/templates/emails/quarterly_stats.html [hunk 8]
```diff
@@ -237,8 +238,52 @@ <h4>An overview of your stats from the last 3 months.</h4>
         <div class="button-action">
             <a class="button" href="{% url 'dashboard' %}">{% trans "Share my stats" %}</a>
         </div>
-    {% endif %}
+        {% endif %}
```
Comment (mbeacom) [inline app/retail/templates/emails/quarterly_stats.html]: This indent doesn't match the starting `{% if %}` block

File: app/retail/templates/emails/quarterly_stats.html [hunk 9]
```diff
@@ -237,8 +238,52 @@ <h4>An overview of your stats from the last 3 months.</h4>
...
+		    <p>Of the {{ user_funded_bounties_count | intcomma }} issue{%if user_funded_bounties_count > 1 %}s{% endif %} that you funded a bounty for, {{ user_num_funded_fulfilled_bounties | intcomma }} were fulfilled -- that's a {{ user_funded_fulfilled_percentage}}% success rate!</p>
+            </div>
+        </div>
+        <div class="centered-contents stat-box row">
+	   <div class="stat-img total-transaction">
+	      <p class="content">{{ user_funded_bounty_developers | length }} contributors</p>
+	   </div>
+            <div class="stat-contents col-sm-12">
+               <p>The following {{ user_funded_bounty_developers | length }} developers have made contributions to your OSS projects this quarter:
+               {% for developer in user_funded_bounty_developers %}{% if forloop.counter == user_num_funded_bounty_developers %} &amp; {% elif forloop.first %}{% else %}, {% endif %}<a href="https://github.com/{{ developer }}">@{{ developer }}</a>{% endfor %}</p>
```
Comment (mbeacom) [inline app/retail/templates/emails/quarterly_stats.html]: Can we drop these fors and ifs to their own lines for readability?

File: app/retail/templates/emails/quarterly_stats.html [hunk 10]
```diff
@@ -237,8 +238,52 @@ <h4>An overview of your stats from the last 3 months.</h4>
...
 
+    	{% if user_funded_bounties_count %}
+        <h2>You've Been Making Money Moves</h2>
+        <h4>An overview of your OSS project's stats from the last 3 months.</h4>
+        <div class="centered-contents stat-box row">
+            <div class="stat-img full-bordered-circle percent-content col-sm-12">
+                <p class="content">{{ user_funded_fulfilled_percentage}} %</p>
+            </div>
+            <div class="stat-contents col-sm-12">
+		    <p>Of the {{ user_funded_bounties_count | intcomma }} issue{%if user_funded_bounties_count > 1 %}s{% endif %} that you funded a bounty for, {{ user_num_funded_fulfilled_bounties | intcomma }} were fulfilled -- that's a {{ user_funded_fulfilled_percentage}}% success rate!</p>
```
Comment (mbeacom) [inline app/retail/templates/emails/quarterly_stats.html]: You can use django's template tag `pluralize` here

File: app/retail/templates/emails/quarterly_stats.html [hunk 11]
```diff
@@ -237,8 +238,52 @@ <h4>An overview of your stats from the last 3 months.</h4>
...
         </div>
-    {% endif %}
+        {% endif %}
 
+    	{% if user_funded_bounties_count %}
+        <h2>You've Been Making Money Moves</h2>
+        <h4>An overview of your OSS project's stats from the last 3 months.</h4>
+        <div class="centered-contents stat-box row">
+            <div class="stat-img full-bordered-circle percent-content col-sm-12">
+                <p class="content">{{ user_funded_fulfilled_percentage}} %</p>
```
Comment (mbeacom) [inline app/retail/templates/emails/quarterly_stats.html]: Can we switch this to: `{{ user_funded_fulfilled_percentage }}`

File: app/retail/templates/emails/quarterly_stats.html [hunk 12]
```diff
@@ -237,8 +238,52 @@ <h4>An overview of your stats from the last 3 months.</h4>
...
+            </div>
+        </div>
+
+        <div class="button-action">
+            <a class="button" href="{% url 'dashboard' %}">{% trans "Share my stats" %}</a>
+        </div>
+    {% endif %}
+{% endif %}
+{% if not user_active_last_quarter and relevant_bounties %}
+    	<h2>You could have been BUILDing</h2>
```
Comment (mbeacom) [inline app/retail/templates/emails/quarterly_stats.html]: This and remaining strings throughout this file need wrapped in `trans` or `blocktrans` (if a variable exists in the string)

File: app/dashboard/models.py [hunk 13]
```diff
@@ -1127,22 +1127,100 @@ def get_quarterly_stats(self):
...
+            bounty for bounty in bounties if bounty.is_funder(self.handle)
+        ]
+        funded_bounties_count = len(funded_bounties)
+        if funded_bounties_count:
+            total_funded_usd = sum([
+                bounty.value_in_usdt if bounty.value_in_usdt else 0
+                for bounty in funded_bounties
+            ])
+            total_funded_hourly_rate = float(0)
+            #TODO: Explain to the user that we only counted Bounties that developers entered # of hours worked on
```
[the remark this answers is not in the document]
Comment (mbeacom) [reply app/dashboard/models.py]: Do you need us to finish the BE for these TODOs?

[the remark this answers is not in the document]
Comment (owocki) [reply app/dashboard/models.py]: i just resolved this TODO

File: app/dashboard/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (oogetyboogety) [reply app/dashboard/models.py]: There is no recommendation on https://www.python.org/dev/peps/pep-0257/, could you advise

File: app/dashboard/models.py (hunk 1 above)
Comment (oogetyboogety) [reply app/dashboard/models.py]: Added TODOs

**Comment by oogetyboogety** [comment]
@owocki @mbeacom Fixed most of the review comments, can you look over once more

**Comment by mbeacom** [comment]
@oogetyboogety Would you mind posting screenshots or gifs for these changes?

**Comment by owocki** [comment]
@oogetyboogety  we are keen to move fowrward on this since its now Q3! let us know what you think

**Comment by oogetyboogety** [comment]
@owocki I'll get those wrapped up and uploaded today!

**Comment by oogetyboogety** [comment]
owocki_rinkeby3
owocki_rinkeby2
owocki_rinkeby
gitcointestdeveloper2

**Comment by owocki** [comment]
these look really good!

**Comment by owocki** [comment]
@oogetyboogety can u pls request payout via the gitcoin bounty?

**Pull request merged**